### ISSUE TITLE
Making the link typing clearer

### ISSUE THREAD

**Issue by dret** [json-api/json-api#1076]
it seems that the language in #document-links is not quite clear when it comes to describing that a link is always typed, and that the link relation type is the member name. in the examples with "self" and "related", it is not explicitly mentioned that these are registered link relations and thus can be used as member name.
to make that even clearer, maybe having a URI as a link relation type in an example would make clear that the "link model" is based on typed links, and that these typed links follow the model established in RFC 5988.

**Comment by wingfire** [comment]
#684 might cover this

**Comment by ethanresnick** [comment]
Yes, #684 should clear this up. The link typing isn't clearer in the spec now because the link keys _aren't_ technically RFC-5988 typed links at the moment, but just named links whose names happen to match common IANA rels.

**Comment by dret** [comment]
that sounds a bit worrying. if they're not RFC 5988, then what's the relationship? isn't RFC 5988 link typing something that pretty much everybody would expect when seeing this kind of "almost RFC 5988" linking? if you need a machine-readable list, maybe http://webconcepts.info/concepts/http-headers can help?

**Comment by ethanresnick** [comment]
> that sounds a bit worrying. if they're not RFC 5988, then what's the relationship?

The answer is that there's almost certainly going to be a mapping to RFC 5988 eventually, and all the keys currently allowed by the specification in `links` objects have meanings compatible with the IANA role of the same name. (Note that users aren't currently allowed to add their own keys to `links` objects either.)

We just haven't said officially what the relationship to RFC 5988 is because there are some details to work out. 

The trouble is caused by some rather disjointed decisions around links in V1 of the spec. For example, each key in `links` is currently defined to hold a single link object rather than an array, which would force links relations to have cardinality of 1 if the key were simply taken to be the relation name. Moreover, the keys currently allowed actually have JSON:API-specific meanings in addition to/that are more narrow than their IANA meanings. So a `links` object like:

``` js
{ "related": "http://example.com/"}
```

implicitly expresses two relations: 
1. {Context IRI} has a "related" link to "http://example.com/"; and
2. {Context IRI} has an "http://jsonapi.org/rels/related" link to "http://example.com/"

A solution to both of these problems (i.e., single link cardinality and implicit external relations) might be to treat the the key as a non-standard name of some sort, with rels expressed in the link object. So `{ "related": {"href": "http://example.com/"} }` would just be sugar for:

``` js
{ 
  "related": {
    "href": "http://example.com/", "rels": ["related", "http://jsonapi.org/rels/related"] 
  } 
}
```

But then we'd want to think about what the key names actually represent, and what we'd want to do if the spec wanted to add keys in the future. It's all solvable, but just takes a bit of thinking we haven't had time for.

There's also the issue that each of our "resource objects" has a key called "relationships", that defines a set of named relationships. These names are just strings at the moment, and are API-specific but, conceptually, they're an awful lot like external link relations. So, in more radical moments I've been noodling around with trying to unify links and relationships...but that would also take some more thinking.

**Comment by dret** [comment]
On 2016-09-19 20:43, Ethan wrote:

> There's also the issue that each of our "resource objects" has a key
> called "relationships", that defines a set of named relationships. These
> names are just strings at the moment, and are API-specific but,
> conceptually, they're an awful lot like external link relations. So, in
> more radical moments I've been noodling around with trying to unify
> links and relationships...but that would also take some more thinking.

i'd be more than happy to help with sorting this out. i think it would 
be very unfortunate to create a rift and make it hard to see. either use 
RFC 5988 and stick to it, or use something else and make that very clear.

if there's anything i can help with on the conceptual side, let me know 
(maybe in a new issue). i am very interested in webby hypermedia and how 
to better support it, so i'd be willing to put in some cycles for 
brainstorming and thinking through different design options.

**Comment by ethanresnick** [comment]
Thanks @dret! That's an awesome offer, and I'll definitely take you up on it when we start really thinking about these issues! It'll hopefully be soon, though there are a couple of non-hypermedia things that we're trying to address first.

Your work on the profile relation has been really influential to me, btw, as I've thought about extensions for JSON:API. (The extension design on the current site is way out of date, but you can see in some [previous discussions](https://github.com/json-api/json-api/issues/915) how direct the influence has been.)

**Comment by dret** [comment]
great to see you're using profiles! thanks, and feel free to reach out when linking becomes a more central issue.
